sim/src/main/scala/bridges/BlockDevModule.scala

```diff
@@ -0,0 +1,67 @@
+//See LICENSE for license details.
+
+package firesim.bridges
+
+import chisel3._
+import chisel3.util._
+
+import firesim.midasexamples.PeekPokeMidasExampleHarness
+import freechips.rocketchip.config.Parameters
+import testchipip.{BlockDeviceData, BlockDeviceIO, BlockDeviceInfo, BlockDeviceRequest}
+
+class BlockDevDUT(implicit val p: Parameters) extends Module {
+  val rd = Module(new BlockDevBridge)
+  rd.io.clock := clock
+  rd.io.reset := reset
+  val rdev = rd.io.bdev
+
+  val wr = Module(new BlockDevBridge)
+  wr.io.clock := clock
+  wr.io.reset := reset
+  val wdev = wr.io.bdev
+
+  // Set up the read and write ends to the same sector.
+  val sector = RegInit(0.U(32.W))
+
+  rdev.req.bits.write  := false.B
+  rdev.req.bits.offset := sector
+  rdev.req.bits.len    := 1.U
+  rdev.req.bits.tag    := 0.U
+
+  wdev.req.bits.write  := true.B
+  wdev.req.bits.offset := sector
+  wdev.req.bits.len    := 1.U
+  wdev.req.bits.tag    := 0.U
+
+  val rd_offset = RegInit(0.U(6.W))
+  val wr_offset = RegInit(0.U(6.W))
+  val waiting   = RegInit(false.B)
+
+  // Initiate a rd when a sector was fully copied.
+  val valid = rd_offset === 0.U && wr_offset === 0.U && sector < rdev.info.nsectors && !waiting
+  rdev.req.valid := valid
```

Comment:
I'm not quite sure how this works correctly since the request side couples req.valid, but the response side decouples it. Is it not possible for the rd and wr offset to get out-of-sync if the block devices back pressure at different times?

It would be more hygenic and (probably) more correct to decouple the enqueue side as well.

Reply:
I tried to schedule requests differently. Hope it's more reasonable now. If something more complex is required, then the FireSim team should do it at a suitable point.